ethereum/core/src/main/java/tech/pegasys/teku/core/ForkChoiceUtil.java

```diff
@@ -235,41 +265,54 @@ public static BlockImportResult on_block(
   }
 
   private static Optional<BlockImportResult> checkOnBlockConditions(
-      final BeaconBlock block, final BeaconState preState, final ReadOnlyStore store) {
+      final BeaconBlock block,
+      final BeaconState preState,
+      final ReadOnlyStore store,
+      final ForkChoiceStrategy forkChoiceStrategy) {
+    final UnsignedLong blockSlot = block.getSlot();
     if (preState == null) {
       return Optional.of(BlockImportResult.FAILED_UNKNOWN_PARENT);
     }
-    if (preState.getSlot().compareTo(block.getSlot()) >= 0) {
+    if (preState.getSlot().compareTo(blockSlot) >= 0) {
       return Optional.of(BlockImportResult.FAILED_INVALID_ANCESTRY);
     }
-    if (blockIsFromFuture(block, store)) {
+    if (blockIsFromFuture(store, blockSlot)) {
       return Optional.of(BlockImportResult.FAILED_BLOCK_IS_FROM_FUTURE);
     }
-    if (!blockDescendsFromLatestFinalizedBlock(block, store)) {
+    if (!blockDescendsFromLatestFinalizedBlock(block, store, forkChoiceStrategy)) {
       return Optional.of(BlockImportResult.FAILED_INVALID_ANCESTRY);
     }
     return Optional.empty();
   }
 
-  private static boolean blockIsFromFuture(BeaconBlock block, ReadOnlyStore store) {
-    return get_current_slot(store).compareTo(block.getSlot()) < 0;
+  private static boolean blockIsFromFuture(ReadOnlyStore store, final UnsignedLong blockSlot) {
+    return get_current_slot(store).compareTo(blockSlot) < 0;
   }
 
   private static boolean blockDescendsFromLatestFinalizedBlock(
-      final BeaconBlock block, final ReadOnlyStore store) {
+      final BeaconBlock block,
+      final ReadOnlyStore store,
+      final ForkChoiceStrategy forkChoiceStrategy) {
     final Checkpoint finalizedCheckpoint = store.getFinalizedCheckpoint();
+    // TODO: throw specific error here
+    final UnsignedLong blockSlot = block.getSlot();
 
     // Make sure this block's slot is after the latest finalized slot
     final UnsignedLong finalizedEpochStartSlot = finalizedCheckpoint.getEpochStartSlot();
-    if (block.getSlot().compareTo(finalizedEpochStartSlot) <= 0) {
+    if (blockSlot.compareTo(finalizedEpochStartSlot) <= 0) {
       return false;
     }
 
     // Make sure this block descends from the finalized block
-    final UnsignedLong finalizedSlot = store.getBlock(finalizedCheckpoint.getRoot()).getSlot();
-    final Bytes32 ancestorAtFinalizedSlot =
-        get_ancestor(store, block.getParent_root(), finalizedSlot);
-    return ancestorAtFinalizedSlot.equals(finalizedCheckpoint.getRoot());
+    final UnsignedLong finalizedSlot =
+        forkChoiceStrategy.blockSlot(finalizedCheckpoint.getRoot()).orElseThrow();
+    try {
+      final Bytes32 ancestorAtFinalizedSlot =
+          get_ancestor(forkChoiceStrategy, block.getParent_root(), finalizedSlot);
+      return ancestorAtFinalizedSlot.equals(finalizedCheckpoint.getRoot());
+    } catch (UnknownBlockRootException e) {
+      return false;
+    }
```

Comment:
```suggestion
   return get_ancestor(forkChoiceStrategy, block.getParent_root(), finalizedSlot)
        .map(
            ancestorAtFinalizedSlot ->
                ancestorAtFinalizedSlot.equals(finalizedCheckpoint.getRoot()))
        .orElse(false);
```

I'm feeling like there's a pattern here so we could just extract a `hasAncestorAtSlot(BeaconBlock block, UnsignedLong slot, Bytes32 expectedBlockRoot)` method.

Reply:
Extracted the method.